### ISSUE TITLE
Variable-length tuple doesn't narrow in `isinstance`

### ISSUE THREAD

**Issue by A5rocks** [python/mypy#15443]
**Bug Report**

This may be a duplicate but I couldn't find any.

See the title for a short description or the short reproduction attached.

**To Reproduce**

```python
a: object
b: tuple[type[int], ...]

if isinstance(a, b):
    reveal_type(a)
```

**Expected Behavior**

```
main.py:5: note: Revealed type is "builtins.int"
Success: no issues found in 1 source file
```

**Actual Behavior**

```
main.py:5: note: Revealed type is "builtins.object"
Success: no issues found in 1 source file
```

This works, however:
```py
a: object
b: tuple[type[int]]

if isinstance(a, b):
    reveal_type(a)
```

**Your Environment**

https://mypy-play.net/?mypy=latest&python=3.11&gist=050eb94523510217a3bf91e9977539d0

**Comment by sobolevn** [comment]
Yes, looks like a problem that can be easily fixed.

Here's a more real-life example that can  illustrate the same problem:

```python
class Base: ...

class Impl1(Base): ...
class Impl2(Base): ...

impls: tuple[type[Base], ...] = (Impl1, Impl2)
some: object

if isinstance(some, impls):
    reveal_type(some)  # must be `Base`
```

I have a fix ready :)
